

# A target_frame for which a transform must exist at the current time before 
# the filter_chain will be executed. This is the target_frame internally passed
# to the tf::MessageFilter. If this parameter is not set, the chain will simply
# be executed immediately upon the arrival of each new scan.
tf_message_filter_target_frame: /bender/sensors/laser_front_link

# Filtering pipeline
# A list of laser filters to load.

scan_filter_chain:

# This filter removes all measurements from the sensor_msgs/LaserScan which
# are greater than upper_threshold or less than lower_threshold. These points
# are "removed" by setting the corresponding range value to max_range + 1,
# which is assumed to be an error case. 
- name: range
  type: laser_filters/LaserScanRangeFilter
  params:
    # Lower range threshold 
    lower_threshold: 0.3
    # Upper range threshold
    upper_threshold: .inf

# This is a filter which removes points inside a polygon, which is obtained as
# a geometry_msgs::PolygonStamped from a ROS topic.
# It was designed to filter points colliding with the robot arms.
- name: robot_self_arms_filter
  type: uchile_laser_pipeline/LaserScanSelfFilter
  params:
    target_frames: [ "bender/l_int_finger_link", "bender/r_int_finger_link" ]
    inflation_radius_list: [ 0.20, 0.20 ]

# # This filter removes laser readings that are most likely caused by the
# # veiling effect when the edge of an object is being scanned. For any two
# # points $$p_1$$ and $$p_2$$, we do this by computing the perpendicular 
# # angle. That is, assuming the origin of the laser is $$O$$, the angle 
# # formed $$\angle OP_1P_2$$. If the perpendicular angle is less than a
# # particular min or greater than a particular max, we remove all neighbors
# # further away than that point. 
# - name: shadows
#   type: laser_filters/ScanShadowsFilter
#   params:
#     # Minimum perpendicular angle 
#     min_angle: 10
#     # Maximum perpendicular angle 
#     max_angle: 170
#     # Number of further-away neighbors to remove
#     neighbors: 20
#     # Number of consecutive measurements to consider angles inside of 
#     window: 1


# # For any measurement in the scan which is invalid, the interpolation comes
# # up with a measurement which is an interpolation between the surrounding 
# # good values. 
# - name: interpolation
#  type: laser_filters/InterpolationFilter




# # This filter internally makes use of the the filters implementation of 
# # float-array filters. It extracts the range and intensity values and treats
# # each as an independent float array passed through an internal filter chain. 
# - name: laser_median_filter
#   type: laser_filters/LaserArrayFilter
#   params: 
#     # A nested filter chain description, describing an appropriate chain
#     # of MultiChannelMedianFilterFloat type filters. 
#     range_filter_chain:
#       - name: median_5
#         type: MultiChannelMedianFilterFloat 
#         params:
#           number_of_observations: 5
#           unused: 10
#     # A nested filter chain description, describing an appropriate chain
#     # of MultiChannelMedianFilterFloat type filters. 
#     intensity_filter_chain:
#       - name: median_5
#         type: MultiChannelMedianFilterFloat 
#         params:
#           number_of_observations: 5
#           unused: 10


# # This filter removes all measurements from the sensor_msgs/LaserScan which are
# # greater than max_intensity or less than min_intensity. These points are 
# # "removed" by setting the corresponding range value to max_range + 1, which is
# # assumed to be an error case.
# - name: intensity
#   type: laser_filters/LaserScanIntensityFilter
#   params:
#     # Intensity value below which readings will be dropped. 
#     lower_threshold: 8000
#     # Intensity value above which readings will be dropped.
#     upper_threshold: 100000
#     # Whether or not to write an intensity histogram to standard out. 
#     disp_histogram: 0


# # This filter removes all measurements from the sensor_msgs/LaserScan which
# # are greater than upper_threshold or less than lower_threshold. These points
# # are "removed" by setting the corresponding range value to max_range + 1,
# # which is assumed to be an error case. 
# - name: range
#   type: laser_filters/LaserScanRangeFilter
#   params:
#     # Lower range threshold 
#     lower_threshold: 0.3
#     # Upper range threshold 
#     upper_threshold: .inf

